spec/lang/step.md

```diff
@@ -554,16 +554,21 @@ impl<M: Memory> Machine<M> {
     ) -> NdResult {
         // First evaluate return place (left-to-right evaluation).
         let ret_place = ret_expr.try_map(|ret_expr| self.eval_place(ret_expr))?;
+        let place_ty = ret_expr.map(
+            |ret_expr| ret_expr.check_wf::<M>(self.cur_frame().func.locals, self.prog).unwrap() // FIXME avoid a second traversal of `ret_expr`
+        );
 
         // Evaluate all arguments.
         let arguments = arguments.try_map(|arg| self.eval_value(arg))?;
 
-        let (value, ret_type) = self.eval_intrinsic(intrinsic, arguments)?;
+        let (value, ret_ty) = self.eval_intrinsic(intrinsic, arguments)?;
 
         if let Some(ret_place) = ret_place {
-            let align = Align::max_for_offset(ret_type.size::<M>()).unwrap();
-            
-            let bytes = Type::encode::<M>(ret_type, value);
+            let place_ty = place_ty.unwrap();
```

Comment:
I think you can avoid the unwrap if you do the ret_expr.check_wf inside the `if let` here.

Reply:
Hmm but ret_expr would still need to be unwrapped. the only thing I would see is to just do `if let Some= ret_expr`. but then `eval_place` happens after evaluating the arguments.

Reply:
Sure, but this `map` outside the `if` should not be needed.